urla134d66615_fikr4n_itl_java_tgz-pJ8-example_ExampleJ8/src/org/arabeyes/itl/prayer/astro/AstroLib.java

```diff
@@ -0,0 +1,491 @@
+			epsilon += (PE[i][2] + JC * PE[i][3])
+			* Math.cos(Utils.degToRad(xsum));
+			xsum = 0;
+		}
+		
+		deltaPsi = psi / 36000000.0;
+		deltaEps = epsilon / 36000000.0;
+		
+		U = JM / 10.0;
+		E0 = 84381.448 - 4680.93 * U - 1.55 * Math.pow(U, 2) + 1999.25
+		* Math.pow(U, 3) - 51.38 * Math.pow(U, 4) - 249.67
+		* Math.pow(U, 5) - 39.05 * Math.pow(U, 6) + 7.12
+		* Math.pow(U, 7) + 27.87 * Math.pow(U, 8) + 5.79
+		* Math.pow(U, 9) + 2.45 * Math.pow(U, 10);
+		E = E0 / 3600.0 + deltaEps;
+		lamda = G + deltaPsi + (-20.4898 / (3600.0 * R));
+		
+		V0 = 280.46061837 + 360.98564736629 * (JD - 2451545) + 0.000387933
+		* Math.pow(JC, 2) - Math.pow(JC, 3) / 38710000.0;
+		V = limitAngle(V0) + deltaPsi * Math.cos(Utils.degToRad(E));
+		
+		RAn = Math.sin(Utils.degToRad(lamda)) * Math.cos(Utils.degToRad(E))
+		- Math.tan(Utils.degToRad(Gg))
+		* Math.sin(Utils.degToRad(E));
+		RAd = Math.cos(Utils.degToRad(lamda));
+		RA = limitAngle(Utils.radToDeg(Math.atan2(RAn, RAd)));
+		
+		DEC = Math.asin(Math.sin(Utils.degToRad(Gg))
+				* Math.cos(Utils.degToRad(E))
+				+ Math.cos(Utils.degToRad(Gg))
+				* Math.sin(Utils.degToRad(E))
+				* Math.sin(Utils.degToRad(lamda)));
+		
+		astroday.setRa(RA);
+		astroday.setDec(DEC);
+		astroday.setSidtime(V);
+		astroday.setDra(0);
+		astroday.setRsum(R);
+		
+	}
+	
+	public static void computeTopAstro(final Location loc, final Astro astro,
+			Astro topAstro) {
+		int i;
+		double lHour, SP;
+		double tU, tCos, tSin, tRA0, tRA, tDEC;
+		
+		for (i = 0; i < 3; i++) {
+			lHour = limitAngle(astro.getSid()[i] + loc.getDegreeLong()
+					- astro.getRa()[i]);
+			
+			SP = 8.794 / (3600 * astro.getRsum()[i]);
+			
+			tU = Math.atan(0.99664719 * Math.tan(Utils.degToRad(loc
+					.getDegreeLat())));
+			
+			tCos = Math.cos(tU) + ((loc.getSeaLevel()) / 6378140.0)
+			* Math.cos(Utils.degToRad(loc.getDegreeLat()));
+			
+			tSin = 0.99664719 * Math.sin(tU) + (loc.getSeaLevel() / 6378140.0)
+			* Math.sin(Utils.degToRad(loc.getDegreeLat()));
+			
+			tRA0 = (((-tCos) * Math.sin(Utils.degToRad(SP)) * Math.sin(Utils
+					.degToRad(lHour))) / (Math.cos(astro.getDec()[i]) - tCos
+							* Math.sin(Utils.degToRad(SP))
+							* Math.cos(Utils.degToRad(lHour))));
+			
+			tRA = astro.getRa()[i] + Utils.radToDeg(tRA0);
+			
+			tDEC = Utils.radToDeg(Math.atan2(
+					(Math.sin(astro.getDec()[i]) - tSin
+							* Math.sin(Utils.degToRad(SP)))
+							* Math.cos(tRA0), Math.cos(astro.getDec()[i])
+							- tCos * Math.sin(Utils.degToRad(SP))
+							* Math.cos(Utils.degToRad(lHour))));
+			
+			topAstro.getRa()[i] = tRA;
+			topAstro.getDec()[i] = tDEC;
+			topAstro.getSid()[i] = astro.getSid()[i];
+			topAstro.getDra()[i] = tRA0;
+			topAstro.getRsum()[i] = astro.getRsum()[i];
+			
+		}
+		
+	}
+	
+	public static double limitAngle(double L) {
+		double F;
+		L /= 360.0;
+		F = L - Math.floor(L);
+		if (F > 0)
+			return 360 * F;
+		else if (F < 0)
+			return 360 - 360 * F;
+		else
+			return L;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 450:457, 462:469.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.